### DISCUSSION TITLE
Heartbeat sender blocked

### DISCUSSION THREAD

**Discussion by danijel-resc** [php-amqplib/php-amqplib#1083] [Help wanted]
In the example `amqp_consumer_pcntl_heartbeat.php`, the consumer uses `sleep`, which is a non-blocking call, as the signals can interrupt it.

If I replace the `sleep` with a `for` loop which lasts for i.e. 3 minutes, then I start getting "missed heartbeat" errors.

What is the correct/recommended way of handling this?

For more context, I'm working on a legacy project that uses Swoole. Once the parent process starts, it opens the connections to the RabbitMQ and waits for the message. Once the message arrives, it will spin up a Swoole process, wait for the process to finish execution, then acknowledge the message. In most cases, the message gets executed in a few seconds (1-6), but we do have a case where the process can take 30 minutes or something. We didn't have this issue with the previous infrastructure, as the connection was just there, forever, and we didn't use heartbeats. However, I'm trying to move these new services to kubernetes, and kubernetes doesn't like long-lasting connections, so I am forced to use heartbeats. Otherwise, we just get "Broken pipe" once it tries to acknowledge the message if the job took a bit longer to complete.

I've tried both the `PCNTLHeartbeatSender` and `SIGHeartbeatSender`. The behavior is the same. I am confident that the reason is the blocking, so the sender can't send heartbeats. I've taken the code from `amqp_consumer_pcntl_heartbeat.php` and replaced the `sleep` call with a `for` loop that loops 100.000.000 times. This takes quite some time and the same thing happens; it throws a "missed heartbeat" error once it reaches the end and tries to acknowledge the message.

I did search for a similar [discussion](https://github.com/php-amqplib/php-amqplib/discussions/1004), but there was no resolution  as the issue was that the sender was in the wrong place.

**Comment by danijel-resc** [comment]
So, since I've posted this, I've tried to find a workaround using Swoole Coroutine and Swoole Process. What I've tried is to have heartbeat sender registration in another child process. This doesn't work for some reason, I guess because of the context. Since the connection is created in one process, then "passed" down to the child process. This results in a "broken pipe" error.

I, then, tried to do the same thing using Swoole Coroutine. No luck with that one either.

I, then, tried to use `swoole_timer_tick`, every 4.000 ms and send the heartbeat manually using `$connection->checkHeartbeat()`, in a child process. Weirdly enough, heartbeats are working this way until the worker gets the message. The child process that runs the job gets spawned, starts working on the job, and then hangs. But the heartbeats are still sent. The job starts hanging right after the first heartbeat which occurs after the job start. I do not understand why, yet.

I am out of ideas at the moment.

**Comment by lukebakken** [comment]
@danijel-resc just FYI I have been on vacation and will try to find time to follow up this coming week. Thanks for your patience and your investigation!

**Comment by danijel-resc** [comment]
@lukebakken I've replaced the whole worker stack with NodeJS. I wasn't the one who decided to use Swoole process manager to execute CLI commands, but it was an awful idea anyway. As I was expecting, the NodeJS AMQP works normally and sends heartbeats even when the CLI command takes some time to finish.

I am still interested to know if there is a workaround for this issue. I literally can not understand how am I the only one with this problem.

The conclusion I came to is that, in the official example, `sleep` actually can stop execution so that the process can react to the system signal, in this case, the `SIGALRM`, which then sends a heartbeat. But, the PHP CLI command can not stop the execution, hence the main process can not react to the `SIGALRM` and send the heartbeat.

**Reply by lukebakken** [reply]
>  As I was expecting, the NodeJS AMQP works normally and sends heartbeats even when the CLI command takes some time to finish

Great, thanks for continuing to use RabbitMQ. I appreciate the update.

My guess is that almost everyone who uses this library does so in the context of a web application, not the CLI.

**Reply by danijel-resc** [reply]
Technically, it's a web application. It's a Symfony application, and the Swoole was used to execute Symfony CLI commands. I guess the previous developers decided to use Swoole simply because it had a Web Socket server. Still, that doesn't change the connection to RabbitMQ.

But I understand your point. This library is awesome when you need communication between two services using queues. The processing of the message doesn't take that long, so it works really well.

Rabbit was never the problem, I love Rabbit! 👍 

If there will be no further debugging, you can close this one then. Thanks!
